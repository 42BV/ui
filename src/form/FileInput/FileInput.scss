.file-upload {
  position: relative;
  &:hover {
    &.form-control-file {
      cursor: pointer;
    }
  }
  &.form-group input[type=file] {
    opacity: 0;
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 100;
    &:hover {
      cursor: pointer;
    }
  }
}
