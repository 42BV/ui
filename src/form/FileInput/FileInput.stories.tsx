import React from 'react';
import { action } from '@storybook/addon-actions';

import {
  FieldFileInput,
  FileInput,
  JarbFileInput,
  requireFile
} from './FileInput';

import {
  FieldFormElementDependencies,
  FinalForm,
  JarbFormElementDependencies
} from '../../story-utils';
import { Alert } from 'reactstrap';
import { Card } from '../../core/Card/Card';
import { Tooltip } from '../../core/Tooltip/Tooltip';
import { Icon } from '../../core/Icon';

export default {
  title: 'Form/FileInput',

  decorators: [
    (Story) => (
      <>
        <Alert color="warning" className="mb-4">
          <p>
            To be able to use FileInput, you have to add lodash to your
            dependencies:
          </p>
          <code>npm install --save lodash</code>
        </Alert>
        <Story />
      </>
    )
  ],

  parameters: {
    component: FileInput
  }
};

const BasicStory = () => {
  return (
    <Card className="m-2">
      <FileInput
        id="file-upload-with-button"
        placeholder="Upload a file here"
        label="Upload a file here"
        accept="text/plain"
        onChange={() => action('value changed')}
      />
    </Card>
  );
};

export const Basic = {
  render: BasicStory,
  name: 'basic'
};

const WithoutPlaceholderStory = () => {
  return (
    <Card className="m-2">
      <FileInput
        id="file-upload-with-button"
        label="Upload a file here"
        accept="text/plain"
        onChange={() => action('value changed')}
      />
    </Card>
  );
};

export const WithoutPlaceholder = {
  render: WithoutPlaceholderStory,
  name: 'without placeholder'
};

const InvisibleLabelStory = () => {
  return (
    <Card className="m-2">
      <FileInput
        id="file-upload-with-button"
        label="Upload a file here"
        hiddenLabel={true}
        placeholder="Upload a file here"
        accept="text/plain"
        onChange={() => action('value changed')}
      />
    </Card>
  );
};

export const InvisibleLabel = {
  render: InvisibleLabelStory,
  name: 'invisible label'
};

const WithCustomLabelStory = () => {
  return (
    <Card className="m-2">
      <FileInput
        id="file-upload-with-button"
        placeholder="Upload a file here"
        label={
          <div className="d-flex justify-content-between">
            <span>Upload a file here</span>
            <Tooltip
              className="ms-1"
              style={{ zIndex: 101 }}
              content="The file should be a plain text file"
            >
              <Icon icon="info" />
            </Tooltip>
          </div>
        }
        accept="text/plain"
        onChange={() => action('value changed')}
      />
    </Card>
  );
};

export const WithCustomLabel = {
  render: WithCustomLabelStory,
  name: 'with custom label'
};

const FieldStory = () => {
  return (
    <>
      <FieldFormElementDependencies />
      <FinalForm>
        <FieldFileInput
          id="file-upload-with-button"
          name="upload"
          placeholder="Upload a file here"
          label="Upload a file here"
          accept="text/plain"
          validators={[requireFile('Profile photo')]}
        />
      </FinalForm>
    </>
  );
};

export const Field = {
  render: FieldStory,
  name: 'field'
};

const JarbStory = () => {
  return (
    <>
      <JarbFormElementDependencies />
      <FinalForm>
        <JarbFileInput
          id="file-upload-with-button"
          name="upload"
          placeholder="Upload a file here"
          accept="text/plain"
          validators={[requireFile('Profile photo')]}
          jarb={{
            validator: 'User.profile',
            label: 'Profile photo'
          }}
        />
      </FinalForm>
    </>
  );
};

export const Jarb = {
  render: JarbStory,
  name: 'jarb'
};
