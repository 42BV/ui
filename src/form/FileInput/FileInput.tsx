import React, { useRef } from 'react';
import { get, noop, uniqueId } from 'lodash';
import { FieldValidator } from 'final-form';

import { FormGroup, Input, InputGroup, Label } from 'reactstrap';

import { withJarb } from '../withJarb/withJarb';
import { doBlur } from '../utils';
import { Translation } from '../../utilities/translation/translator';
import { AddonButton } from '../addons/AddonButton/AddonButton';
import { FieldCompatible } from '../types';
import { Icon } from '../../core/Icon';

type Props = FieldCompatible<File, File | null> & {
  /**
   * Which types of files the form element accepts.
   *
   * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#Unique_file_type_specifiers
   */
  accept: string;
};

export function FileInput(props: Props) {
  const {
    id = uniqueId(),
    label,
    accept,
    placeholder,
    error,
    color,
    value,
    valid,
    className = '',
    onBlur,
    onFocus
  } = props;

  const inputRef = useRef<HTMLInputElement|null>(null);

  function onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const files = event.target.files;

    /* istanbul ignore else */
    if (files) {
      const file = files.item(0);

      props.onChange(file);

      doBlur(onBlur);
    }
  }

  function onClick(event: React.MouseEvent<HTMLInputElement, MouseEvent>) {
    // The 'delete' button is clicked in this case.
    if (value) {
      event.preventDefault();

      props.onChange(null);
      doBlur(onBlur);

      if (inputRef.current) {
        inputRef.current.value = '';
      }
    }

    // The input should now trigger the browsers file dialog
  }


  const name = get(value, 'name', '');
  const inputProps = {
    placeholder,
    value: name,
    invalid: valid === false ? true : undefined,
    onChange: () => console.error('The placeholder input of FileInput should not be changeable')
  };

  return (
    <FormGroup className={`file-upload ${className}`} color={color}>
      {label ? <Label for={id}>{label}</Label> : null}
      <input
        id={id}
        accept={accept}
        ref={inputRef}
        type="file"
        onClick={onClick}
        onChange={onChange}
        onFocus={onFocus}
      />

      <InputGroup>
        <Input {...inputProps} />
        <AddonButton
          position="right"
          color={value || valid === false ? 'danger' : 'primary'}
          onClick={noop}
        >
          <Icon icon={value ? 'delete' : 'cloud_upload'} className="d-block" />
        </AddonButton>
      </InputGroup>

      {error}
    </FormGroup>
  );
}

/**
 * Variant of the FileInput which can be used in a Jarb context.
 */
export const JarbFileInput = withJarb<File, File | null, Props>(FileInput);

/**
 * A FileValidator is a FieldValidator which checks if the `File`
 * is valid.
 */
export type FileValidator = FieldValidator<File>;

/**
 * Takes a `label` and returns a validator which can check if the
 * there is a `File`.
 *
 * @export
 * @param {string} label The label to display in the error message
 * @returns {(FieldValidator<File>)}
 */
export function requireFile(label: string): FileValidator {
  return (value: File | null | undefined): Translation | undefined => {
    if (value) {
      return undefined;
    }

    return {
      key: 'FileInput.REQUIRED',
      data: { label },
      fallback: `${label} is required`
    };
  };
}

/**
 * Takes a `size` and `label` and returns a validator which can check if the
 * file is past the size limit.
 *
 * @export
 * @param {string} size The maximum size of the file in MB.
 * @param {string} label The label to display in the error message
 * @returns {(FieldValidator<File | null | undefined>)}
 */
export function limitFileSize(size: number, label: string): FileValidator {
  return (value: File | null | undefined): Translation | undefined => {
    if (!(value instanceof File)) {
      return undefined;
    }

    // Older browsers will not support this so we skip the check.
    /* istanbul ignore if */
    if (value.size === undefined) {
      return undefined;
    }

    const maxSizeInBytes = size * 1048576;

    if (value.size < maxSizeInBytes) {
      return undefined;
    }

    const fileSize = (value.size / 1048576).toFixed(1);
    const maxSizeDisplay = size.toFixed(1);

    return {
      key: 'FileInput.SIZE_TOO_LARGE',
      data: { label, size: maxSizeDisplay, fileSize },
      fallback: `${label} file is to large. Max size is ${maxSizeDisplay} MB file size is ${fileSize} MB`
    };
  };
}
