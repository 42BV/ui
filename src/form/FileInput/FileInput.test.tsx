import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import { FileInput, limitFileSize, requireFile } from './FileInput';

describe('Component: FileInput', () => {
  function setup({
    value,
    valid,
    hasPlaceholder = true,
    hasLabel = true
  }: {
    value?: File;
    valid?: boolean;
    hasPlaceholder?: boolean;
    hasLabel?: boolean;
  }) {
    const onChangeSpy = vi.fn();
    const onBlurSpy = vi.fn();

    const props = {
      placeholder: hasPlaceholder ? 'Upload a file here' : undefined,
      accept: 'text/plain',
      value,
      onChange: onChangeSpy,
      onBlur: onBlurSpy,
      error: 'Some error',
      valid,
      id: hasLabel ? 'file-upload-with-button' : undefined,
      label: 'Upload a file here',
      hiddenLabel: !hasLabel
    };

    const { container } = render(<FileInput color="success" {...props} />);

    return { container, onChangeSpy, onBlurSpy };
  }

  describe('ui', () => {
    test('with value', () => {
      const { container } = setup({
        value: new File([''], 'maarten.png'),
        valid: true
      });
      expect(container).toMatchSnapshot();
    });

    test('without value', () => {
      setup({ hasLabel: true });
      // @ts-expect-error Form elements have property value
      expect(screen.getByLabelText('Upload a file here').value).toEqual('');
    });

    test('without placeholder', () => {
      setup({ hasPlaceholder: false });
      expect(screen.queryByPlaceholderText('Upload a file here')).toBeNull();
    });

    test('invisible label', async () => {
      expect.assertions(1);
      setup({ hasLabel: false });
      expect(screen.queryByText('Upload a file here')).toBeNull();
      await screen.findByLabelText('Upload a file here');
    });

    test('invalid', () => {
      setup({ valid: false });
      expect(
        screen
          .queryByPlaceholderText('Upload a file here')
          ?.classList.contains('is-invalid')
      ).toBe(true);
    });

    test('valid', () => {
      setup({ valid: true });
      expect(
        screen
          .queryByPlaceholderText('Upload a file here')
          ?.classList.contains('is-invalid')
      ).toBe(false);
    });
  });

  test('should not be able to change placeholder input', () => {
    const console = global.console;
    const consoleErrorSpy = vi.fn();
    // @ts-expect-error We only use error, not all the other properties
    global.console = { error: consoleErrorSpy };

    setup({});

    fireEvent.change(screen.getByPlaceholderText('Upload a file here'), {
      target: { value: 'test' }
    });

    expect(consoleErrorSpy).toHaveBeenCalledTimes(1);
    expect(consoleErrorSpy).toHaveBeenCalledWith(
      'The placeholder input of FileInput should not be changeable'
    );

    global.console = console;
  });

  describe('events', () => {
    test('onChange', () => {
      const { onChangeSpy, onBlurSpy } = setup({ value: undefined });

      const file = new File([''], 'maarten.png');

      fireEvent.change(screen.getByLabelText('Upload a file here'), {
        target: { files: { item: vi.fn().mockReturnValue(file) } }
      });

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy.mock.calls[0][0].name).toBe('maarten.png');

      expect(onBlurSpy).toHaveBeenCalledTimes(1);
    });

    describe('onClick', () => {
      test('with value', () => {
        const { onChangeSpy, onBlurSpy } = setup({
          value: new File([''], 'maarten.png')
        });

        fireEvent.click(screen.getByLabelText('Upload a file here'));

        expect(onChangeSpy).toHaveBeenCalledTimes(1);
        expect(onChangeSpy).toHaveBeenCalledWith(null);

        expect(onBlurSpy).toHaveBeenCalledTimes(1);

        // @ts-expect-error Form elements have property value
        expect(screen.getByLabelText('Upload a file here').value).toEqual('');
      });

      test('without value', () => {
        const { onChangeSpy, onBlurSpy } = setup({ value: undefined });

        fireEvent.click(screen.getByLabelText('Upload a file here'));

        expect(onChangeSpy).toHaveBeenCalledTimes(0);
        expect(onBlurSpy).toHaveBeenCalledTimes(0);
      });
    });
  });
});

test('requireFile', () => {
  const validator = requireFile('cv');

  // @ts-expect-error Even though it accepts File it will be given undefined and null
  expect(validator(undefined, {})).toEqual({
    data: { label: 'cv' },
    key: 'FileInput.REQUIRED',
    fallback: 'cv is required'
  });
  // @ts-expect-error Even though it accepts File it will be given undefined and null
  expect(validator(null, {})).toEqual({
    data: { label: 'cv' },
    key: 'FileInput.REQUIRED',
    fallback: 'cv is required'
  });

  expect(validator(new File([''], 'henkie.docx'), {})).toBe(undefined);
});

test('limitFileSize', () => {
  const validator = limitFileSize(1, 'cv');

  // @ts-expect-error Even though it accepts File it will be given undefined and null
  expect(validator(undefined, {})).toBe(undefined);

  // @ts-expect-error Even though it accepts File it will be given undefined and null
  expect(validator(null, {})).toBe(undefined);

  const smallFile = new File([''], 'small.docx');
  expect(validator(smallFile, {})).toBe(undefined);

  const largeFile = new File(['very long string'.repeat(100000)], 'large.docx');
  expect(validator(largeFile, {})).toEqual({
    data: { fileSize: '1.5', label: 'cv', size: '1.0' },
    fallback: 'cv file is to large. Max size is 1.0 MB file size is 1.5 MB',
    key: 'FileInput.SIZE_TOO_LARGE'
  });
});
