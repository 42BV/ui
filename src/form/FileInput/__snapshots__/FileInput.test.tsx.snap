// Vitest Snapshot v1

exports[`Component: FileInput > ui > with value 1`] = `
<div>
  <div
    class="file-upload  mb-3"
    color="success"
  >
    <label
      class="form-label"
      for="file-upload-with-button"
    >
      Upload a file here
    </label>
    <input
      accept="text/plain"
      id="file-upload-with-button"
      type="file"
    />
    <div
      class="input-group"
    >
      <input
        aria-label="This is a placeholder input that should not be reachable. Use the file input instead."
        class="form-control"
        defaultvalue="maarten.png"
        placeholder="Upload a file here"
        tabindex="-1"
        type="text"
      />
      <span
        class="button d-inline-block danger"
      >
        <button
          class="btn btn-danger btn-md"
          type="button"
        >
          <div
            class="d-flex justify-content-center align-items-center"
          >
            <i
              class="icon d-block material-icons"
            >
              delete
            </i>
          </div>
        </button>
      </span>
    </div>
    Some error
  </div>
</div>
`;
